Adding greet feature to chatbot when tabs are switched

amannaik247 opened the pull request:
Problem: When you switch to the chatbot tab for the first time it greets you with ["Hi I am robot (name). Ask me any question"]. 
This is a useful audio indication, since there are no visual changes from the ABC tab to chatbot tab.
But when a user changes tabs back to chatbot the second time(or more) they don't get any audio cue. 

Solution: I have added a small audio cue ["Hi again! I'm (name)"], everytime a user changes their tab to the chabot.

This audio cue will be a nice addition since it is gives a clear indication when user changes to the chatbot section.

MostlyKIGuess commented on brain.py:
```diff
@@ -114,26 +111,28 @@ def load_brain():
             logger.debug('Load bot: %s' % brain)
 
-            kernel = Kernel()
+            # Only create new kernel if voice changed or no kernel exists
```
so you removed lines 96-98 to add this check here which is redundant if you hadn't removed them in the first place,

MostlyKIGuess commented on brain.py:
```diff
@@ -146,8 +145,11 @@ def load_brain():
+            if sorry:
+                activity.face.say_notification(sorry)
+            else:
+                activity.face.say_notification(_("Hi again! I'm %s.") % brain_name)
```
so this hi again is the only viable change and there are 5 commits which is weird

amannaik247 replied:
Since I have changed the flow of the same code with some additions . Same code is on different lines, which is why you are seeing these changes.
I am following the specific format of any other activity.
I hope that clears this doubt.

MostlyKIGuess reviewed (changes requested):
So maybe just add the last check because everything else is just already implemented and rebase it using
```sh
git rebase -i HEAD~5
```

5-> number of commits here. You can squash all of them into one.

amannaik247 commented:
@MostlyKIGuess 
I have rebased the branch to have just one commit message.
Removed the check line as well.

amannaik247 commented:
@chimosky please review this PR and the changes made. Thank you!

quozl commented:
* [ ] rewrite your commit message based on our guidelines,
* [ ] remove the unnecessary changes.

Thanks!

amannaik247 commented:
> * [ ]  rewrite your commit message based on our guidelines,
> * [ ]  remove the unnecessary changes.
> 
> Thanks!

Rewrote the commit message and removed any unnecessary changes.

chimosky commented:
> all of this is again not needed

I also agree with this comment from MostlyK, did you try to figure out why that was happening? 
What was the intended behaviour and does the code do that?

amannaik247 commented:
Originally:
 If the voice hasn't changed, the function stops completely and no greeting is played. Due to this condition(line 96):
```
    if voice == _kernel_voice:
        return False
```
Changed code:
If the voice hasn't changed, the function still runs (allowing the greeting to play), but we skip recreating the kernel. 

Therefore, I moved the condition because we want the greeting to play every time, but we only want to recreate the kernel when necessary.

The intended behaviour of the code is to have the bot greet the user everytime they switch back to the chatbot (giving audio indication).
Which was achieved. Please check the attached video. 

https://github.com/user-attachments/assets/31ea125e-ea9e-4716-9d38-48a27c20148d

amannaik247 commented:
> No, it does not. See https://github.com/sugarlabs/sugar-docs/blob/master/src/contributing.md#making-commits

Rewrote the commit message, accordingly. I hope this is follows the guidelines.

quozl commented:
I've been in classrooms where children have chosen Speak activity, and the greeting is one of the most least pleasant aspects of the activity.  Making it even more frequent doesn't sound well thought out.  I'm fine with the other things being fixed, but unprompted sound should be kept to a minimum.

chimosky commented:
@amannaik247 said;
> If the voice hasn't changed, the function still runs (allowing the greeting to play), but we skip recreating the kernel.
>
> Therefore, I moved the condition because we want the greeting to play every time, but we only want to recreate the kernel when necessary.
>
> The intended behaviour of the code is to have the bot greet the user everytime they switch back to the chatbot (giving audio indication).

Removing the check solves the problem, why do we need the extra changes?

@quozl said;
> I've been in classrooms where children have chosen Speak activity, and the greeting is one of the least pleasant aspects of the activity. Making it even more frequent doesn't sound well thought out. I'm fine with the other things being fixed, but unprompted sound should be kept to a minimum.

I'm fine with the greeting playing every time the bot gets focus because kids these days might enjoy it.

amannaik247 commented:
> I've been in classrooms where children have chosen Speak activity, and the greeting is one of the least pleasant aspects of the activity. Making it even more frequent doesn't sound well thought out. I'm fine with the other things being fixed, but unprompted sound should be kept to a minimum.

That is why I have kept the audio cue very small. So that it doesn't bother much.

amannaik247 commented:
> Removing the check solves the problem, why do we need the extra changes?

@chimosky Yes, removing the check does solve the problem. But every time we switch back to the chatbot it reloads the brain(slowing the process).

This is a screenshot of how the brain gets loaded 4 times when I switch back to it 4 times[When I only remove the check]
![image](https://github.com/user-attachments/assets/b841a9d3-9034-4eb8-9424-3515821b8962)

But in the end when i switch back to my code where I have placed the condition at a different place. The brain only loads once. It will only load again if the voice is changed. But still plays the greeting every time.

quozl commented:
Thanks.  That wasn't clear from your commit message; could you please
* [ ] separate the change into two commits, one for your added feature, and one for reducing the reload,
* [ ] write into the commit messages a more complete description of why the change is being done,
* [ ] make sure the wrapping is fixed,

Currently the single commit message looks like this;

```log
	The early return condition for voice check was preventing the
	greeting
	from playing when switching tabs. Moving this check to only
	control
	kernel creation allows the greeting to play while still
	maintaining
	efficient kernel management.

	The function now plays the greeting message every time user
	switches
	to the chatbot tab, while only recreating the kernel when
	necessary.
```

Which is hard to read because one word is split onto next line.

Please wrap the text to within 70 columns, as `git` on command line is typically used in an 80 column terminal.

MostlyKIGuess commented:
> ![image](https://private-user-images.githubusercontent.com/168978808/437957104-b841a9d3-9034-4eb8-9424-3515821b8962.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDU4MTkxMTEsIm5iZiI6MTc0NTgxODgxMSwicGF0aCI6Ii8xNjg5Nzg4MDgvNDM3OTU3MTA0LWI4NDFhOWQzLTkwMzQtNGViOC05NDI0LTM1MTU4MjFiODk2Mi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNDI4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDQyOFQwNTQwMTFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0wZWRmMzliYmY2YTg0ZGE1M2NjNTc4NzU1ZmVmMGY3ZGI0YzdkOWYzOTE0NTI3MzcyNjEzM2M2YjI3NzAyYzUyJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.vAzsZOOpLOTR6zUp2ggrrZhnsFnG-ase4xl-FYMwfEE)

This is a really nice improvement! I assume you are testing the time when brain loads.

Interesting to note that time for the app still remains almost the same.

Master:
sugar-activity3  0.88s user 0.11s system 45% cpu 2.152 total
This PR:
sugar-activity3  0.92s user 0.12s system 44% cpu 2.352 total

Although my system isn't the best to test this, as I didn't use a VM to test it where it would be used. But interesting overall!

amannaik247 commented:
> Thanks. That wasn't clear from your commit message; could you please
> 
> * [x]  separate the change into two commits, one for your added feature, and one for reducing the reload,
> * [x]  write into the commit messages a more complete description of why the change is being done,
> * [x]  make sure the wrapping is fixed,

- Made 2 separate commits, each with a proper commit message and description of the changes made.
- Ensured the wrapping is fixed. 

@chimosky please review. Thank you for the guidance!

chimosky commented on brain.py:
```diff
@@ -148,6 +145,8 @@ def load_brain():
 
     GLib.idle_add(load_brain)
-    return True
+    return True
```
Please remove the no new line at EOF, you can do so with `echo >> brain.py`, take a look at git diffs before making commits, it'll help with preventing issues like this.

chimosky commented:
I'm curious why your commit message body is spaced the way it is, why the extra space at the beginning?

quozl commented:
I agree.  Remove the spaces at the start of line.  I know I could do it myself, but I don't want to have to keep doing it.   Whitespace is your responsibility.  Also fix typo chabot.

I've looked at the history of commits to brain.py, and the very first commit contains the rather strange design of `load_brain` inside `load`.  load as a function name is overloaded.  There's also very little point doing this in an idle task; all that achieves is complete redraw before the hang of loading.  Perhaps it should be a background thread.

Having re-read the history, and having been there with testing back in 2009, I remember now that the intention of the "hello" was to indicate reinitialisation of the brain model after several seconds, and for automatic responses to be minimised, so that the child learner perceives the model as obedient and polite.  I still don't think the user should be greeted each time, whereas @chimosky thinks it should happen because it might be enjoyed.

With initialisation time now so small, is it worth deferring at all?

amannaik247 commented:
> I'm curious why your commit message body is spaced the way it is, why the extra space at the beginning?

Removed the extra spacing in the commit message body 

> Having re-read the history, and having been there with testing back in 2009, I remember now that the intention of the "hello" was to indicate reinitialisation of the brain model after several seconds, and for automatic responses to be minimised, so that the child learner perceives the model as obedient and polite. I still don't think the user should be greeted each time, whereas @chimosky thinks it should happen because it might be enjoyed.

Thank you for sharing your perspective and the original intention. I agree that politeness and not overwhelming the learner are important. However, a brief, friendly greeting when the learner returns to the tab can create a welcoming atmosphere and reinforce engagement. As long as the greeting is short, it can make the experience feel more interactive and enjoyable without being intrusive.
Therefore, I have shortened the greeting to just "Hi again!".

quozl commented:
I disagree with the added greeting, but two against one.  Proceeding.

* [ ] brain.py has lost the end of the last line after one commit, then gets it back again in the next commit, why?

chimosky commented:
> I disagree with the added greeting, but two against one. Proceeding.
> 
>     * [ ]  brain.py has lost the end of the last line after one commit, then gets it back again in the next commit, why?

Typical removal of No new line at EOF without checking the diff to notice it, the second commit is adding it back after I mentioned it in the first.

quozl commented:
> Typical removal of No new line at EOF without checking the diff to notice it, the second commit is adding it back after I mentioned it in the first.

Heh.  Yes, I know.  I was using a rhetorical device to elicit reflection.

quozl commented:
Well done.

chimosky commented:
> Heh. Yes, I know. I was using a rhetorical device to elicit reflection.

Hehe, should've known.
